From 991b03be3f6f2af221c6e350f782fd8cefcf3fce Mon Sep 17 00:00:00 2001
From: Yongzhi Chen <yongzhi_chen@hotmail.com>
Date: Sun, 25 Oct 2015 23:24:31 -0700
Subject: [PATCH 0363/1164] CDH-33419 HIVE-12189: The list in pushdownPreds of
 ppd.ExprWalkerInfo should not be allowed to grow
 very large (Yongzhi Chen, reviewed by Chao Sun)

Change-Id: I5e119c77ada29e612fa065231307b7d5a16d72e8
---
 .../apache/hadoop/hive/ql/ppd/ExprWalkerInfo.java  |   23 ++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/ppd/ExprWalkerInfo.java b/ql/src/java/org/apache/hadoop/hive/ql/ppd/ExprWalkerInfo.java
index 33ad3e8..3b8a2e5 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/ppd/ExprWalkerInfo.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/ppd/ExprWalkerInfo.java
@@ -228,7 +228,13 @@ public void addFinalCandidate(ExprNodeDesc expr) {
   }
 
   public void addFinalCandidate(String alias, ExprNodeDesc expr) {
-    getPushdownPreds(alias).add(expr);
+    List<ExprNodeDesc> predicates = getPushdownPreds(alias);
+    for (ExprNodeDesc curPred: predicates) {
+      if (curPred.isSame(expr)) {
+        return;
+      }
+    }
+    predicates.add(expr);
   }
 
   /**
@@ -238,7 +244,20 @@ public void addFinalCandidate(String alias, ExprNodeDesc expr) {
    * @param pushDowns
    */
   public void addPushDowns(String alias, List<ExprNodeDesc> pushDowns) {
-    getPushdownPreds(alias).addAll(pushDowns);
+    List<ExprNodeDesc> predicates = getPushdownPreds(alias);
+    boolean isNew;
+    for (ExprNodeDesc newPred: pushDowns) {
+      isNew = true;
+      for (ExprNodeDesc curPred: predicates) {
+        if (curPred.isSame(newPred)) {
+          isNew = false;
+          break;
+        }
+      }
+      if (isNew) {
+        predicates.add(newPred);
+      }
+    }
   }
 
   /**
-- 
1.7.9.5

